Tab completion of  variable expression based on the value of the variable

powercode opened the pull request:
Fixes #2567 

Using the value of a variable as a last resort when trying to infer the type of a variable expression.

msftclas commented:
Hi __@powercode__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

lzybkr reviewed (changes requested):
I'm not keen on this change for a few reasons:
- `GetInferredType` is inferring, this change is somewhat more precise
- `GetInferredType` shouldn't require a runspace
- This change only handles simple variables. A real fix should handle more complicated expressions.

I think the fix should be in `System.Management.Automation.CompletionCompleters.NativeCompletionMemberName` and it should call `SafeExprEvaluator.TrySafeEval` to get the value of the expression.

powercode commented:
Cool. Reworking it.

Commit:
Adding support for getting completion based on the value of a variable.

$p = Get-Process
$p | foreach <Tab>
$p[0] | foreach <Tab>
$p.Where{$true} | sort description |
      where BasePriority -eq 13 | % <Tab>
now works.

powercode commented:
New attempt. Eager to learn about better ways to do this.

$p = gps
$p | % <tab>
$p[0] | % <tab> 
works
$p.Where{$true} | % <Tab> does not.

powercode commented:
@lzybkr I wonder if the right thing to do here is the give the CompletionContext a property TrySafeEval, and have CommandExpressionAst check if it should try a safeeval when inferring it's type.

lzybkr commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -3793,9 +3793,70 @@ private static void NativeCompletionMemberName(string wordToComplete, List<Compl
+        static bool IsCompletionTypeInvariantCommand(CommandAst commandAst)
+        {
+            var commandName = commandAst.CommandElements[0].Extent.Text;
+            if (string.Compare(commandName, "where", StringComparison.OrdinalIgnoreCase) == 0) return true;
```
For faster comparisons, you should use the string that most likely appears in the script, which is usually whatever tab completion returns, so `Where-Object` instead of `where-object`.

lzybkr commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -3793,9 +3793,70 @@ private static void NativeCompletionMemberName(string wordToComplete, List<Compl
             CompleteMemberByInferredType(context, prevType, result, wordToComplete + "*", filter: IsPropertyMember, isStatic: false);
+            if (result.Count == 0)
+            {
+                // if we get here, we have not been able to infer the type to complete.
```
Most of this code feels like it should be in it's own method because it might be useful elsewhere, basically everything up to the call to CompleteMemberByActualType.

lzybkr reviewed (changes requested):
Can you add some tests?

We don't have many tab completion tests ported out of our internal framework yet, but there are a few if you search for `TabExpansion2` and `CompletInput`.

That said, if you feel like introducing some cleaner infrastructure to make it easy to add tab completion tests, that would be greatly appreciated.

powercode commented:
I wonder if the key here is 

> GetInferredType shouldn't **require** a runspace

I think it will be impossible to cover a lot of scenarios if `GetInferredType `can't ever use a runspace.
I understand that there are situations when this is necessary, but it will limit what can be deduced by type inference.

If the `GenInferredType `methods could conditionally use a runspace (determined by the context passed in) we could have the best of both worlds.

VariableExpressions, InvokeMemberExpressions, IndexExpression and probably others all depend on the possibility to query the current runspace.

@lzybkr How do you feel about a property on the context, or a parameter on GetInferredType to indicate if runspace use is allowed when inferring type?

Otherwise it will be hard to not end up with lots of special cases that has to be solved in hacky ways.

My PR cannot handle for example 
```powershell
$p = Get-Process
$p | Foreach {$_.MainModule} | Foreach Com<Tab>
```
because it is not using the type inference, it just bypasses it for a common special case.

lzybkr commented:
I'm all for a robust api that works for both static and dynamic determination of the type of an Ast, ideally one that is robust enough to make public (which `GetInferredType` is not - it works well enough, but it takes some shortcuts which can be completely wrong at times,)

`GetInferredType` is already too tightly coupled to completion - and I'd like to see it move in the other direction if possible to make it more generally useful, e.g. for PSScriptAnalyzer.

One way to start is to move `GetInferredType` out of `Ast` so that the `Ast` isn't as tightly coupled to the rest of the PowerShell engine. Some tools could benefit from a parser decoupled from the rest of the PowerShell engine.

powercode commented:
I agree. The dependency from the AST to CompletionContext is kind of strange even though that was a convenient place to put it. 
Is type inference generally useful? I.E. in other places than completion?

lzybkr commented:
Script analyzer also needs type inference. You can imagine wanting to report a warning on members that don't exist (a likely typo).

powercode commented:
Is it a class TypeInferer with overloads for each AST? 
Where each method is moved and transformed to take the AST as the first parameter?
```
class TypeInferer{
   public TypeInferer(SomeContextOtherThanCompletionContext context)
   public IEnumerable<PSTypeName> GetInferedType(VariableExpressionAst ast)...
   public IEnumerable<PSTypeName> GetInferedType(InvokeMemberExpressionAst ast)...
   ...
}
```

lzybkr commented:
We use the GoF `Visitor` pattern to "add" virtual methods to the `Ast`.

You can see many examples if you look at implementations of `ICustomAstVisitor2`. Note that it's often useful to simply derive from `DefaultCustomAstVisitor2` so you can skip implementing all of the methods.

Note that `ICustomAstVisitor2` is very different than `AstVisitor2` which actually walks the tree for you, calling you for each node in the Ast.

powercode commented:
I will close this pull request, and open a new one with an implementation based on what @lzybkr  suggested, i.e. a ICustomAstVistitor2 based implementation.
